### PR TITLE
feat(go): runtime type casting

### PR DESCRIPTION
**Pull request by MrArnoldPalmer**
Adds an implementation map to all code generated getters and method
invocations to pass to the runtime. This allows the runtime to construct
concrete types when the return type of the function is a go interface.

Adds additional tests in `jsii-calc-test` module for various use cases.
Adds logic to correctly deserialize and serialize object references
coming from both the host and kernel.

fix: #2047 

---

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license].

[Apache 2.0 license]: https://www.apache.org/licenses/LICENSE-2.0

### CODE REVIEW

**Review by SoManyHs — COMMENTED** [review]
Looks good to me, some non-blocking comments on the tests but otherwise 👍 

Would be great if @RomainMuller could take a gander as well!

File: packages/jsii-pacmak/lib/targets/go/package.ts
Comment (MrArnoldPalmer) [reply packages/jsii-pacmak/lib/targets/go/package.ts:97]: Will defer to your method once its merged.

File: packages/@jsii/go-runtime/jsii-calc-test/main_test.go
Comment (MrArnoldPalmer) [reply packages/@jsii/go-runtime/jsii-calc-test/main_test.go:57]: definitely can/should/will be.

File: packages/@jsii/go-runtime/jsii-calc-test/main_test.go
Comment (MrArnoldPalmer) [reply packages/@jsii/go-runtime/jsii-calc-test/main_test.go:66]: We aren't necessarily asserting that. but that is the return type of the method so thats the type that is usable to the user. The test is mostly 'Returns an interface{} but actually has a string that can be extracted via reflection'. I'll change the description here.

File: packages/@jsii/go-runtime/jsii-calc-test/main_test.go
Comment (MrArnoldPalmer) [reply packages/@jsii/go-runtime/jsii-calc-test/main_test.go:73]: Yeah these are funky because as far as I can tell its the only way to do a [type matching](https://tour.golang.org/methods/16) without reflection. It seems to be the preferred way of handling this.

File: packages/@jsii/go-runtime/jsii-calc-test/main_test.go
Comment (MrArnoldPalmer) [reply packages/@jsii/go-runtime/jsii-calc-test/main_test.go:26]: ooh I like that

Will work on this separately.

File: packages/@jsii/go-runtime/jsii-experimental/runtime.go
Comment (RomainMuller) [inline packages/@jsii/go-runtime/jsii-experimental/runtime.go:11]: That could have used a documentation string explaining what the key versus value is supposed to be.

File: packages/jsii-pacmak/lib/targets/go/runtime/method-call.ts
Comment (RomainMuller) [inline packages/jsii-pacmak/lib/targets/go/runtime/method-call.ts:78]: Looks like this would be invoked a bunch of times and could be kinda slow & thrash's... Maybe store the result in a private and re-use?

File: packages/jsii-pacmak/lib/targets/go/runtime/method-call.ts
Comment (RomainMuller) [inline packages/jsii-pacmak/lib/targets/go/runtime/method-call.ts:24]: You don't seem to ever add things to this map (same below on L48)... And it really just gets passed to a method...

This makes me wonder:
- Do we need the variable, in fact?
- Can we not use an "empty map literal" instead of using `make` here?

File: packages/jsii-pacmak/lib/targets/go/runtime/property-access.ts
Comment (RomainMuller) [inline packages/jsii-pacmak/lib/targets/go/runtime/property-access.ts:29]: Same as before - is there any reason we cannot use a literal here?

And since this map is fed with respects to `this.parent.reference?`, can we not have this be computed *once* and stored somewhere for re-use?

File: packages/jsii-pacmak/lib/targets/go/runtime/property-access.ts
Comment (RomainMuller) [inline packages/jsii-pacmak/lib/targets/go/runtime/property-access.ts:28]: (Also below on line 68)

It would be nice to have a comment on this bad boy showing what a line it produces could look like... The amount of braces there is nauseating and it's kinda hard to picture what this produces.

File: packages/jsii-pacmak/lib/targets/go/types/class.ts
Comment (RomainMuller) [inline packages/jsii-pacmak/lib/targets/go/types/class.ts:253]: Just double-not things to coerce them into a boolean context...

```suggestion
    return !!this.reference?.scopedImplMap(this.parent.pkg).length;
```

File: packages/jsii-pacmak/lib/targets/go/types/class.ts
Comment (RomainMuller) [inline packages/jsii-pacmak/lib/targets/go/types/class.ts:252]:
```suggestion
  public get usesReflectionPackage(): boolean {
```

File: packages/jsii-pacmak/lib/targets/go/types/go-type-reference.ts
Comment (RomainMuller) [inline packages/jsii-pacmak/lib/targets/go/types/go-type-reference.ts:90]: That looks like something that perhaps should be stored (it's used in 3 distinct locations...).

File: packages/jsii-pacmak/lib/targets/go/types/go-type-reference.ts
Comment (RomainMuller) [inline packages/jsii-pacmak/lib/targets/go/types/go-type-reference.ts:117]:
```suggestion
  public scopedImplMap(scope: Package): [string, string] {
```

Also - this returns a pair of strings, which isn't extremely expressive. Perhaps returning some object `{ scopedInterface: string, scopedImplementingType: string }` would be more readable?

File: packages/jsii-pacmak/lib/targets/go/types/go-type-reference.ts
Comment (RomainMuller) [inline packages/jsii-pacmak/lib/targets/go/types/go-type-reference.ts:140]: Optional (I'm even starting to doubt myself here) - I kinda find moving the bang after the parentheses makes it better signal "one of these is defined"... I don't know...

```suggestion
          (typeMap.array ?? typeMap.map)!.typeMap,
```

File: packages/jsii-pacmak/lib/targets/go/types/go-type-reference.ts
Comment (RomainMuller) [inline packages/jsii-pacmak/lib/targets/go/types/go-type-reference.ts:172]: This this even possible/legal? Should we throw here?

File: packages/jsii-pacmak/lib/targets/go/types/go-type-reference.ts
Comment (RomainMuller) [inline packages/jsii-pacmak/lib/targets/go/types/go-type-reference.ts:183]: Since `scopedTypeName` will actually only care about ONE of them at a time, it appears to me this type is excessively permissive (it'll allow all the fields to be set, which appears to be illegal).

I'd also make this as immutable as possible to make sure nothing does anything stupid with that.

```suggestion
type TypeMap =
     { readonly primitive?: string; }
  || { readonly array?: GoTypeRef; }
  || { readonly map?: GoTypeRef; }
  || { readonly union?: readonly GoTypeRef[]; }
  || { readonly void?: boolean; }
  || { readonly interface?: GoTypeRef; }
```

**Review by RomainMuller — COMMENTED** [review]
I don't think I've seen an `implMap` value have more than 1 entry in it (they're either empty, or have exactly one key-value pair). Are we over-engineering here? Do we really need a map? Is the fact we never get >1 item just because our tests don't get into the use-case that'd cause > 1 item to be generated (and then - can we fix that?)

**Review by SoManyHs — APPROVED** [review]

**Pull request merged**
